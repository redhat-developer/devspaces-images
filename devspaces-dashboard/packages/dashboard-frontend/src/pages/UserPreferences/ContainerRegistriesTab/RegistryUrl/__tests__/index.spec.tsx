/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { RegistryUrlFormGroup } from '..';

describe('Registry Url Input', () => {
  const mockOnChange = jest.fn();

  function getComponent(url: string): React.ReactElement {
    return (
      <Form>
        <RegistryUrlFormGroup url={url} onChange={mockOnChange} />
      </Form>
    );
  }

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should correctly render the component', () => {
    const component = getComponent('https://ttt');
    render(component);

    const input = screen.queryByRole('textbox');

    expect(input).toBeTruthy();

    const json = renderer.create(component).toJSON();

    expect(json).toMatchSnapshot();
  });

  it('should show default placeholder', () => {
    const component = getComponent('');
    render(component);

    const placeholder = screen.queryByPlaceholderText('Enter a registry');
    expect(placeholder).toBeTruthy();
  });

  it('should correctly re-render the component', () => {
    const component = getComponent('http://test');
    const { rerender } = render(component);

    const input = screen.getByRole('textbox');
    expect(input).toHaveValue('http://test');

    rerender(getComponent('http://testreg.com'));

    expect(input).toHaveValue('http://testreg.com');
  });

  it('should fire onChange event', () => {
    const component = getComponent('http://t');
    render(component);

    const input = screen.getByRole('textbox');

    userEvent.type(input, 'est');

    expect(mockOnChange).toHaveBeenCalledWith('http://te', 'success');
    expect(mockOnChange).toHaveBeenCalledWith('http://tes', 'success');
    expect(mockOnChange).toHaveBeenCalledWith('http://test', 'success');
  });

  describe('validation', () => {
    it('should handle empty value', () => {
      const component = getComponent('https://testreg.com/test1');
      render(component);

      const input = screen.getByRole('textbox');

      userEvent.clear(input);

      const label = screen.queryByText('A value is required.');

      expect(label).toBeTruthy();
      expect(input).toBeInvalid();
    });

    it('should handle maximum value length', () => {
      const component = getComponent('https://testreg.com/test1');
      render(component);

      const input = screen.getByRole('textbox');

      const message = 'The url is too long. The maximum length is 256 characters.';
      const allowedUrl = 'a'.repeat(256);
      let label: HTMLElement | null;

      userEvent.clear(input);
      userEvent.type(input, allowedUrl);

      label = screen.queryByText(message);
      expect(label).toBeFalsy();

      const disallowedUrl = 'a'.repeat(257);

      userEvent.clear(input);
      userEvent.type(input, disallowedUrl);

      label = screen.queryByText(message);
      expect(label).toBeTruthy();
      expect(input).toBeInvalid();
    });
  });
});
