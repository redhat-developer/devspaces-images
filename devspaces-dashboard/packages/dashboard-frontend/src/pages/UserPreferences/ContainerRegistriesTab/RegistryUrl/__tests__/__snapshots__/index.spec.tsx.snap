// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registry Url Input should correctly render the component 1`] = `
<form
  className="pf-c-form"
  noValidate={true}
>
  <div
    className="pf-c-form__group"
    style={
      Object {
        "gridTemplateColumns": "80px",
        "minHeight": "65px",
      }
    }
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="id-registry-helper"
      >
        <span
          className="pf-c-form__label-text"
        >
          Registry
        </span>
        <span
          aria-hidden="true"
          className="pf-c-form__label-required"
        >
           
          *
        </span>
      </label>
       
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <span
        className="pf-c-input-group__text"
      >
        <input
          aria-invalid={false}
          aria-label="Url input"
          className="pf-c-form-control"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Enter a registry"
          readOnly={false}
          required={false}
          type="url"
          value="https://ttt"
        />
        <button
          aria-disabled={true}
          aria-label="open registry"
          className="pf-c-button pf-m-link pf-m-disabled"
          data-ouia-component-id="OUIA-Generated-Button-link-2"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe={true}
          disabled={true}
          role={null}
          tabIndex={null}
          type="button"
        >
          <a
            href="https://ttt"
            rel="noreferrer"
            style={
              Object {
                "color": "inherit",
              }
            }
            target="_blank"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              fill="currentColor"
              height="1em"
              role="img"
              style={
                Object {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
              />
            </svg>
          </a>
        </button>
      </span>
      
    </div>
  </div>
</form>
`;
